p2p, refactor: make NetPermissionFlags an enum class

jonatack opened the pull request:
While reviewing #20196, I noticed the `NetPermissionFlags` enums are frequently called as if they were scoped, yet are still global. This patch upgrades `NetPermissionFlags` to a scoped class enum and updates the enumerator naming, similarly to #19771. See https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#enum-enumerations for more info.

This change would eliminate the class of bugs like https://github.com/bitcoin/bitcoin/pull/20196#discussion_r610770148 and #21644, as only defined operations on the flags would compile.

jonatack committed:
scripted-diff: rename NetPermissionFlags enumerators

- drop redundant PF_ permission flags prefixes
- drop ALL_CAPS naming per https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-caps
- rename IsImplicit to Implicit

-BEGIN VERIFY SCRIPT-
s() { git grep -l "$1" src | xargs sed -i "s/$1/$2/g"; }

s 'PF_NONE'        'None'
s 'PF_BLOOMFILTER' 'BloomFilter'
s 'PF_RELAY'       'Relay'
s 'PF_FORCERELAY'  'ForceRelay'
s 'PF_DOWNLOAD'    'Download'
s 'PF_NOBAN'       'NoBan'
s 'PF_MEMPOOL'     'Mempool'
s 'PF_ADDR'        'Addr'
s 'PF_ISIMPLICIT'  'Implicit'
s 'PF_ALL'         'All'
-END VERIFY SCRIPT-

jonatack committed:
scripted-diff: add NetPermissionFlags scopes where not already present

-BEGIN VERIFY SCRIPT-
s() { git grep -l "$1" -- 'src' ':!src/net_permissions.h' | xargs sed -i -E "s/([^:])$1/\1NetPermissionFlags::$1/"; }

s 'PF_NONE'
s 'PF_BLOOMFILTER'
s 'PF_RELAY'
s 'PF_FORCERELAY'
s 'PF_DOWNLOAD'
s 'PF_NOBAN'
s 'PF_MEMPOOL'
s 'PF_ADDR'
s 'PF_ISIMPLICIT'
s 'PF_ALL'
-END VERIFY SCRIPT-

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

jonatack committed:
p2p, refactor: make NetPermissionFlags a uint32 enum class

and define/update operation methods to handle type conversions explicitly. See
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-oper
for more info.

practicalswift commented:
Strong Concept ACK

[`enum class` is strictly better :)](https://github.com/bitcoin/bitcoin/pull/19771#issuecomment-678733782)

MarcoFalke commented on src/net.cpp:
```diff
@@ -968,7 +968,7 @@ bool CConnman::AttemptToEvictConnection()
         LOCK(cs_vNodes);
         for (const CNode* node : vNodes) {
-            if (node->HasPermission(NetPermissionFlags::PF_NOBAN))
+            if (node->HasPermission(NetPermissionFlags::NoBan))
```
meta: Could the rename be a scripted-diff?

jonatack replied:
done

kiminuo commented on src/test/netbase_tests.cpp:
```diff
@@ -381,41 +381,47 @@ BOOST_AUTO_TEST_CASE(netpermissions_test)
+
+    NetPermissions::ClearFlag(whitebindPermissions.m_flags, NetPermissionFlags::Implicit);
+    BOOST_CHECK(!NetPermissions::HasFlag(whitebindPermissions.m_flags, NetPermissionFlags::Implicit));
+    BOOST_CHECK(!NetPermissions::HasAnyFlag(whitebindPermissions.m_flags, NetPermissionFlags::Implicit));;
```
```suggestion
    BOOST_CHECK(!NetPermissions::HasAnyFlag(whitebindPermissions.m_flags, NetPermissionFlags::Implicit));
```
?

jonatack replied:
good eye, thanks! the compiler didn't mind but we do. fixed.

kiminuo commented on src/net_permissions.h:
```diff
@@ -14,46 +15,55 @@ struct bilingual_str;
+        using t = typename std::underlying_type<NetPermissionFlags>::type;
+        return (static_cast<t>(flags) & static_cast<t>(f)) == static_cast<t>(f);
+    }
+    static inline bool HasAnyFlag(NetPermissionFlags flags, NetPermissionFlags f) {
```
Could you please explain what was the motivation to add `HasAnyFlag`? Your tests for `HasFlag` and `HasAnyFlag` seem to be the same, hence the question.

kiminuo commented:
Concept ACK, nice change

kiminuo commented on src/test/netbase_tests.cpp:
```diff
@@ -427,14 +433,23 @@ BOOST_AUTO_TEST_CASE(netpermissions_test)
+
+    NetPermissions::ClearFlag(whitelistPermissions.m_flags, NetPermissionFlags::Download);
+    BOOST_CHECK(!NetPermissions::HasFlag(whitelistPermissions.m_flags, NetPermissionFlags::Download));
+    BOOST_CHECK(!NetPermissions::HasFlag(whitelistPermissions.m_flags, NetPermissionFlags::NoBan));
```
Maybe move this line one line below? So that it is clear how `HasFlag` and `HasAnyFlag` differ.

jonatack replied:
sure, done

jonatack commented:
> Could you please explain what was the motivation to add HasAnyFlag?

Thanks for the review @kiminuo! I extracted the `NetPermissions::HasAnyFlag()` addition to a separate commit with a detailed explanation in the commit message and updated with your other feedback.

jonatack commented:
rebased

jonatack commented:
rebased

hebasto reviewed (commented):
Approach ACK fa73a18ad874cc076a9201678ff25686766c24d0.

The first commit "p2p, refactor: add NetPermissionFlags scopes where not already present" could be a scripted-diff:
```sh
s() { git grep -l "$1" -- 'src' ':!src/net_permissions.h' | xargs sed -i -E "s/([^:])$1/\1NetPermissionFlags::$1/"; }

s 'PF_NONE'
s 'PF_BLOOMFILTER'
s 'PF_RELAY'
s 'PF_FORCERELAY'
s 'PF_DOWNLOAD'
s 'PF_NOBAN'
s 'PF_MEMPOOL'
s 'PF_ADDR'
s 'PF_ISIMPLICIT'
s 'PF_ALL'
```

hebasto commented on src/net.cpp:
```diff
@@ -2410,7 +2414,7 @@ bool CConnman::Bind(const CService &addr, unsigned int flags, NetPermissionFlags
     }
 
-    if (addr.IsRoutable() && fDiscover && !(flags & BF_DONT_ADVERTISE) && !(permissions & PF_NOBAN)) {
+    if (addr.IsRoutable() && fDiscover && !(flags & BF_DONT_ADVERTISE) && !NetPermissions::HasAnyFlag(permissions, NetPermissionFlags::NoBan)) {
```
Why not `NetPermissions::HasFlag` ?

jonatack replied:
Hm, the tests pass with `HasFlag()` but ISTM that would represent a potential change in behavior in a refactoring pull, as `HasFlag()` returns `(flags & f) == f` instead of `flags & f` and `NetPermissionFlags::NoBan` is a multi-flag (`NoBan = (1U << 4) | Download`). LMK if I'm confused here though.

Aside, this line was last changed in commit bb145c90502:
```diff
-    if (addr.IsRoutable() && fDiscover && (permissions & PF_NOBAN) == 0) {
+    if (addr.IsRoutable() && fDiscover && !(flags & BF_DONT_ADVERTISE) && !(permissions & PF_NOBAN)) {
```
There do seem to be some oddities and unused code WRT the net permissions flags that could be cleaned up in a separate patch.

hebasto replied:
You are right. Your change is equivalent to the current code. But this raises additional questions:

1. current tests are not reliable if they
> pass with `HasFlag()`

2. `HasAnyFlag(permissions, flag_a | flag_b)` is much easier to read and reason about than `HasAnyFlag(permissions, fancy_multiflag)`

3. in general, with named multiflags, negation of `HasFlag` could have ambiguous meaning. Does this code do what it is supposed to do: https://github.com/bitcoin/bitcoin/blob/590e49ccf2af27c6c1f1e0eb8be3a4bf4d92ce8b/src/net.cpp#L1110-L1117 ?

jonatack replied:
After looking at #16248 and #19191, the current code seems correct.

This line was changed in e5b26deaaa6, which pre-dates the download permission being extracted from noban in #19191: 
```diff 
-    if (addrBind.IsRoutable() && fDiscover && !fWhitelisted)
+    if (addrBind.IsRoutable() && fDiscover && (permissions & PF_NOBAN) == 0)
```
The intent of adding the Download flag in #19191 was: *"It should be possible to grant nodes in a local network (e.g. home, university, enterprise, ...) permission to download blocks even after the maxuploadtarget is hit. Currently this is only possible by setting the noban permission, which has some adverse effects, especially if the peers can't be fully trusted. Fix this by extracting a download permission from noban."*

So, in reply to your questions:

1. IIUC the existing lone bitwise AND here, e.g. `HasAnyFlag(Noban)` is correct

2. I guess we could replace `HasAnyFlag(Noban)` with the more verbose `HasFlag(Noban) || HasFlag(Download)` if that seems clearer

3. Elsewhere, `HasFlag()` appears to be what we want in the cases where it is used, AFAICT

MarcoFalke replied:
Ugh, looks like this is a bug that I introduced in #19191

jonatack replied:
I first thought that too, but here in Bind() I'm not sure.

MarcoFalke replied:
Well, fixing a bug should indeed be done in a separate pull from one that is tagged with [refactoring]

jonatack replied:
Ok will open a separate patch for that.

jonatack replied:
Done in #21644. Resolving here and will simplify this PR, dropping the `HasAnyFlags()` and unit test commits.

jonatack commented:
> The first commit "p2p, refactor: add NetPermissionFlags scopes where not already present" could be a scripted-diff:
> 
> ```shell
> s() { git grep -l "$1" -- 'src' ':!src/net_permissions.h' | xargs sed -i -E "s/([^:])$1/\1NetPermissionFlags::$1/"; }

Thanks! That's a co-credit :) Done in 043a4c5163909d0.

jonatack committed:
scripted-diff: update noban documentation in net_processing.cpp

-BEGIN VERIFY SCRIPT-
s() { git grep -l "$1" src/net_processing.cpp | xargs sed -i "s/$1/$2/g"; }
s 'the noban permission'      'NetPermissionFlags::NoBan permission'
s 'the NOBAN permission flag' 'NetPermissionFlags::NoBan permission'
s 'noban permission'          'NetPermissionFlags::NoBan permission'
-END VERIFY SCRIPT-

jonatack commented:
Simplified, dropped the `HasAnyFlag()` and test commits that are replaced by the bugfix in #21644 and on which this patch is now based.

vasild commented:
There is still the following "strange" behavior:

```cpp
// expected to have "noban" and "download" in flags, true
flags = TryParse("noban,download");

// expected to have just "noban" in flags, given that "download" has been removed, not true :(
ClearFlag(flags, PF_DOWNLOAD);

// what happens actually is that now flags is in a "bricked" state
// where it does not equal any of the PF_* flags or a bitwise-OR of them.
```

Could it be avoided to have the download flag included in the noban flag (i.e. to have each PF_* flag have exactly one bit set) and https://github.com/bitcoin/bitcoin/pull/19191 resolved in some other way?

jonatack commented:
I don't disagree, though it might be out of scope for this change (which has indeed exposed some issues!)

`ClearFlag` is currently only used in `src/net.cpp::CConnman::CreateNodeFromAcceptedSocket()` to clear the NetPermissionFlags implicit flag.

vasild commented on src/net_permissions.h:
```diff
@@ -14,46 +15,51 @@ struct bilingual_str;
+    Implicit = (1U << 31),
+    All = BloomFilter | ForceRelay | Relay | NoBan | Mempool | Download | Addr,
 };
+static inline constexpr NetPermissionFlags operator|(NetPermissionFlags a, NetPermissionFlags b) {
```
nit: the opening `{` should be on a new line

jonatack replied:
done

vasild commented on src/net_permissions.h:
```diff
@@ -14,46 +15,51 @@ struct bilingual_str;
-    static inline bool HasFlag(const NetPermissionFlags& flags, NetPermissionFlags f)
-    {
-        return (flags & f) == f;
+    static inline bool HasFlag(NetPermissionFlags flags, NetPermissionFlags f) {
```
nit: the opening `{` should be on a new line

jonatack replied:
done, thanks!

vasild reviewed (approved):
ACK 24b294ac7519d6989d9746edd9e6643a4adbe88a

jonatack commented:
Thanks @vasild! Will update along with the bracket syntax suggestions once #21644 is sorted out.

Rspigler commented:
Mark up for grabs?

jonatack committed:
p2p: NetPermissions::HasFlag() pass flags param by value

jonatack commented:
Re-opening and rebased now that #21644, that this is based on, was merged.

vasild reviewed (approved):
ACK d53c70e775e56c4e4c30728262121c9f7e4bcc23

laanwj commented on src/net_permissions.h:
```diff
@@ -14,52 +15,59 @@ struct bilingual_str;
     static inline void AddFlag(NetPermissionFlags& flags, NetPermissionFlags f)
     {
-        flags = static_cast<NetPermissionFlags>(flags | f);
+        flags = (flags | f);
```
nit: unnecessary parenthesis

jonatack replied:
fixed

MarcoFalke commented on src/net_permissions.h:
```diff
@@ -14,52 +15,59 @@ struct bilingual_str;
+    Implicit = (1U << 31),
+    All = BloomFilter | ForceRelay | Relay | NoBan | Mempool | Download | Addr,
 };
+static inline constexpr NetPermissionFlags operator|(NetPermissionFlags a, NetPermissionFlags b)
```
I think the goal of this pull is to forbid the `&` operator, because it won't do what you think it does. (See the preceding bugfix)

MarcoFalke replied:
That is, callers should use `HasFlag` instead of `&`.

jonatack replied:
The equivalent of `&` is in `HasFlag()` just below, and we use that function everywhere rather than `&`.

Abstracting out a reusable metatype would be neat.

jonatack replied:
Oh, didn't see @MarcoFalke's replies. Yes.

maflcko reviewed (approved):
> drop ALL_CAPS naming per https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-caps

TIL

MarcoFalke commented:
cr ACK d53c70e775e56c4e4c30728262121c9f7e4bcc23

jonatack commented:
> > drop ALL_CAPS naming per https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-caps

Yes--was wondering if we should update the enum example in our developer notes.

jonatack commented:
Dropped unneeded parentheses, only change is `git diff d53c70e 7075f60`
```diff
diff --git a/src/net_permissions.h b/src/net_permissions.h
index 9cb59bd85f..7a158aa6c5 100644
--- a/src/net_permissions.h
+++ b/src/net_permissions.h
@@ -56,7 +56,7 @@ public:
     }
     static inline void AddFlag(NetPermissionFlags& flags, NetPermissionFlags f)
     {
-        flags = (flags | f);
+        flags = flags | f;
     }
```

vasild reviewed (approved):
ACK 7075f604e8d0b21b2255fa57e20cd365dc10a288

luke-jr reviewed (approved):
utACK

laanwj commented:
Code review ACK 7075f604e8d0b21b2255fa57e20cd365dc10a288
